### PR TITLE
Add support of snap command

### PR DESCRIPTION
**Pull request by vadmeste**
Add snapshot command implement: mc snap -h for full command list.

### CODE REVIEW

File: cmd/snap-serialize.go [hunk 1]
```diff
@@ -399,14 +399,16 @@ func (s *snapshotDeserializer) SkipBucketEntries() *probe.Error {
 // BucketEntries will return all bucket entries.
 // The channel will be closed when there are no more entries.
 // If an error occurs it will be returned and the channel will be closed.
-func (s *snapshotDeserializer) BucketEntries(ctx context.Context, entries chan<- SnapshotEntry) *probe.Error {
```
Comment (klauspost) [inline cmd/snap-serialize.go]: @vadmeste The context provides that functionality.

Comment (vadmeste) [reply cmd/snap-serialize.go]: > @vadmeste The context provides that functionality.

The problem with context is that it always provides an error, but I want from BucketEntries to return nil in that case.

**Comment by vadmeste** [comment]
Not WIP anymore.

**Comment by klauspost** [comment]
@vadmeste Could you merge master? I will add my review.

File: cmd/client-snapshot.go [hunk 2]
```diff
@@ -0,0 +1,582 @@
...
+		APIType: "snapshot",
+	})
+}
+
+// Versioning operations
+
+// Not implemeneted
+func (s *snapClient) GetVersioning(ctx context.Context) (minio.BucketVersioningConfiguration, *probe.Error) {
+	return minio.BucketVersioningConfiguration{}, probe.NewError(APINotImplemented{
+		API:     "GetLifecycle",
```
Comment (klauspost) [inline cmd/client-snapshot.go:545]:
```suggestion
		API:     "GetVersioning",
```
Couldn't this be supported?

File: cmd/client-snapshot.go [hunk 3]
```diff
@@ -0,0 +1,582 @@
...
+		APIType: "snapshot",
+	})
+}
+
+// Not implemented
+func (s *snapClient) SetVersioning(ctx context.Context, status string) *probe.Error {
+	return probe.NewError(APINotImplemented{
+		API:     "SetLifecycle",
```
Comment (klauspost) [inline cmd/client-snapshot.go:553]:
```suggestion
		API:     "SetVersioning",
```

File: cmd/snap-main.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+	"strings"
+
+	"github.com/minio/cli"
+	"github.com/minio/mc/pkg/probe"
+)
+
+// Snapshot command
+var snapCmd = cli.Command{
+	Name:            "snap",
+	Usage:           "generate snapshots of S3 objects",
```
Comment (klauspost) [inline cmd/snap-main.go:31]:
```suggestion
	Usage:           "generate snapshots of S3 object metadata",
```

File: cmd/snap-remove.go [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+	"github.com/minio/mc/pkg/probe"
+	"github.com/minio/minio/pkg/console"
+	"golang.org/x/crypto/ssh/terminal"
+)
+
+var (
+	snapRemoveFlags = []cli.Flag{
+		cli.BoolFlag{
+			Name:  "force, f",
+			Usage: "Force removing the snapshot without a prompt",
```
Comment (klauspost) [inline cmd/snap-remove.go:38]: Do we really need to ask? It is 'only' a snapshot.

Just drop the confirmation IMO.

File: docs/minio-client-complete-guide.md [hunk 6]
```diff
@@ -862,6 +863,128 @@ myobject.txt:    14 B / 14 B  ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓
 Waiting move operations to complete
 ```
 
+<a name="snap"></a>
+### Command `snap`
+
+`snap` command generates snapshots of S3 data at a given date/time. It relies on S3 bucket versioning feature to pick objects versions.
```
Comment (klauspost) [inline docs/minio-client-complete-guide.md:869]:
```suggestion
`snap` command generates snapshots of S3 metadata at a given date/time. It relies on S3 bucket versioning feature to pick objects versions.
```

We do not in any way want to indicate that it stores actual data (for now at least).

Comment (klauspost) [reply docs/minio-client-complete-guide.md:869]: Could you add a sentence on what it can be used for?

**Review by klauspost — COMMENTED** [review]
Nothing big.

**Comment by harshavardhana** [comment]
Are we ready for this now? @vadmeste

File: cmd/snap-remove.go (hunk 5 above)
Comment (vadmeste) [reply cmd/snap-remove.go:38]: > Do we really need to ask? It is 'only' a snapshot.
> 
> Just drop the confirmation IMO.

Let's wait until this is approved/disapproved by AB.

I always want confirmation for delete.

**Comment by vadmeste** [comment]
I rebased this PR and added some fixes. It is ready for review now.

File: cmd/snap-remove.go (hunk 5 above)
Comment (klauspost) [reply cmd/snap-remove.go:38]: `mc rm file` doesn't require confirmation. 

We are not deleting actual data, so I don't really see why it is so important.

**Comment by klauspost** [comment]
@vadmeste Needs rebase

Are we only waiting for review of this one, or has it deliberately been put on hold?

**Comment by klauspost** [comment]
I think we should offer to snapshot file contents as well, but this could rather easily be extended to include that at some point.

File: cmd/client-s3.go [hunk 7]
```diff
@@ -1,5 +1,5 @@
 /*
- * MinIO Client (C) 2015-2020 MinIO, Inc.
+ * MinIO Client (C) 2020 MinIO, Inc.
```
Comment (kannappanr) [inline cmd/client-s3.go:2]:
```suggestion
 * MinIO Client (C) 2020 MinIO, Inc.
```
```suggestion
 * MinIO Client (C) 2015-2020 MinIO, Inc.
```

File: cmd/config.go [hunk 8]
```diff
@@ -307,6 +307,26 @@ func expandAlias(aliasedURL string) (alias string, urlStr string, aliasCfg *alia
 	if aliasCfg = mustGetHostConfig(alias); aliasCfg != nil {
 		return alias, urlJoinPath(aliasCfg.URL, path), aliasCfg, nil
 	}
+
+	snapshots, err := listSnapshots()
```
Comment (kannappanr) [inline cmd/config.go:311]: Please add comments if relevant

**Review by klauspost — APPROVED** [review]

**Comment by harshavardhana** [comment]
Let's revisit this later when we have time - closing for now.

**Pull request closed without merging**